controlplane/internal/controller/openshiftassistedcontrolplane_controller.go

```diff
@@ -210,15 +212,47 @@ func (r *OpenshiftAssistedControlPlaneReconciler) Reconcile(ctx context.Context,
 	result := ctrl.Result{}
 	if conditions.IsTrue(oacp, controlplanev1alpha2.KubeconfigAvailableCondition) {
 		// in case upgrade is still in progress, we want to requeue, however we also want to reconcile replicas
-		result, err = r.upgradeWorkloadCluster(ctx, cluster, oacp, log, pullsecret)
+		result, err = r.upgradeWorkloadCluster(ctx, cluster, oacp, architectures, pullsecret)
 		if err != nil {
 			return result, err
 		}
 	}
 	return result, r.reconcileReplicas(ctx, oacp, cluster)
 }
 
-func (r *OpenshiftAssistedControlPlaneReconciler) upgradeWorkloadCluster(ctx context.Context, cluster *clusterv1.Cluster, oacp *controlplanev1alpha2.OpenshiftAssistedControlPlane, log logr.Logger, pullSecret []byte) (ctrl.Result, error) {
+func getBootstrapConfigArchitectures(ctx context.Context, k8sClient client.Client, oacp *controlplanev1alpha2.OpenshiftAssistedControlPlane) ([]string, error) {
+
+	if oacp == nil {
+		return []string{}, nil
+	}
+	clusterName, ok := oacp.Labels[clusterv1.ClusterNameLabel]
+	if !ok {
+		return []string{}, nil
+	}
+	labelSelector := map[string]string{
+		clusterv1.ClusterNameLabel: clusterName,
+	}
+	listOptions := []client.ListOption{
+		client.InNamespace(oacp.Namespace),
+		client.MatchingLabels(labelSelector),
+	}
+	var configList bootstrapv1alpha1.OpenshiftAssistedConfigList
+	if err := k8sClient.List(ctx, &configList, listOptions...); err != nil {
+		return []string{}, err
+	}
+
+	architectures := make([]string, 0)
+	for _, config := range configList.Items {
+		if config.Spec.CpuArchitecture != "" {
+			architectures = append(architectures, config.Spec.CpuArchitecture)
+		}
+	}
```

Comment:
What do you think about moving the logic from [`getArchitecture`](https://github.com/openshift-assisted/cluster-api-agent/pull/102/files#diff-9bf377e2c7bc277e1ec29fb657fa201dacbe0e599fa8abea4cc1b577425021c6R66-R83) to here so that way we only need to loop through once to get the architecture?

Reply:
Good idea! The code is actually more clear this way. I've decided to loop through them twice for simplicity and better readability. I don't think it's problematic performance wise, as the List operation is the most expensive one, second pass it's only in memory architecture strings comparison. WDYT?

Reply:
Looks great!! 🤩